src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -462,8 +604,50 @@ private void onSignal(
                 {
                     writer.doAbort(target, targetRouteId, targetStreamId, traceId, targetAuthorization);
                 }
+            }
+        }
+
+        private boolean tryExtendAuthStateExpiration(
+            int sourceRealmId)
+        {
+            boolean extendedExpiration = false;
+
+            try
+            {
+                final JwtClaims claims = JwtClaims.parse(signature.getPayload());
+                final String subject = claims.getSubject();
+                final Map<String, OAuthAccessGrant> authStateMap = lookupAuthState(sourceRealmId, subject);
+
+                if(authStateMap != null)
+                {
+                    final OAuthAccessGrant authState = authStateMap.get(subject.intern());
+                    final long delay = authState.expiresAt - System.currentTimeMillis();
+
+                    extendedExpiration = delay >= 0;
 
+                    if(extendedExpiration)
+                    {
+                        this.expiryFuture = executor.schedule(delay, MILLISECONDS,
+                                targetRouteId, targetStreamId, TOKEN_EXPIRED_SIGNAL);
+                    }
+                }
             }
+            catch (InvalidJwtException | JoseException | MalformedClaimException e)
+            {
+                // TODO: diagnostics?
+            }
+            return extendedExpiration;
+        }
+
+        private Map<String, OAuthAccessGrant> lookupAuthState(
+            int sourceRealmId,
+            String subject)
+        {
+            return inFlightAuthorizationsByRealm[sourceRealmId].values()
+                                                               .stream()
+                                                               .filter(map -> map.containsKey(subject.intern()))
+                                                               .findFirst()
```

Comment:
Can this be replaced by `return inFlightAuthorizationsByRealm[sourceRealmId].get(subject.intern());` ?

Reply:
That would have to be `inFlightAuthorizationsByRealm[sourceRealmId].get(affinity).get(subject.intern());` which we wouldn't have unless we store affinity as a private variable or something.

Reply:
Wouldn't it just be another argument to the method, called from a context where we receive the `BEGIN` (which has `affinity` on the frame)?